src/audio/dai-zephyr.c

```diff
@@ -250,13 +252,69 @@ static enum dma_cb_status dai_dma_cb(struct dai_data *dd, struct comp_dev *dev,
 
 	local_buf = buffer_acquire(dd->local_buffer);
 
-	if (dev->direction == SOF_IPC_STREAM_PLAYBACK)
+	if (dev->direction == SOF_IPC_STREAM_PLAYBACK) {
 		ret = dma_buffer_copy_to(local_buf, dma_buf,
 					 dd->process, bytes);
-	else
-		ret = dma_buffer_copy_from(dma_buf, local_buf,
-					   dd->process, bytes);
+		buffer_release(local_buf);
+	} else {
+		struct list_item *sink_list;
+
+		audio_stream_invalidate(&dma_buf->stream, bytes);
+		ret = dma_buffer_copy_from_no_consume(dma_buf, local_buf, dd->process, bytes);
+		buffer_release(local_buf);
+#if CONFIG_IPC_MAJOR_4
+		/* copy from DMA to all sink buffers using the right PCM converter function */
+		list_for_item(sink_list, &dev->bsink_list) {
+			struct comp_buffer __sparse_cache *sink_c;
+			struct comp_dev *sink_dev;
+			struct comp_buffer *sink;
+			int j;
+
+			sink = container_of(sink_list, struct comp_buffer, source_list);
+
+			/* this has been handled above already */
```

Comment:
would it be possible to remove that special case and just handle it in the loop like all others?

Reply:
I dont want to modify the special case because I think there's something off with the converter function. The dd-> process is set during dai_zephyr_params() for the single sink buffer case and I want to keep it that way until I figure out whats wrong with the converter function.